flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesLeaderElector.java

```diff
@@ -86,12 +106,33 @@ public KubernetesLeaderElector(
                                                         newLeader,
                                                         leaderConfig.getConfigMapName())))
                         .build();
-        internalLeaderElector =
-                new LeaderElector(kubernetesClient, leaderElectionConfig, executorService);
+        this.executorService = executorService;
+
         LOG.info(
-                "Create KubernetesLeaderElector {} with lock identity {}.",
-                leaderConfig.getConfigMapName(),
-                leaderConfig.getLockIdentity());
+                "Create KubernetesLeaderElector on lock {}.",
+                leaderElectionConfig.getLock().describe());
+    }
+
+    @GuardedBy("lock")
+    private void resetInternalLeaderElector() {
+        cancelCurrentLeaderElectionSession();
+
+        currentLeaderElectionSession =
+                currentLeaderElectionSession.thenCompose(
```

Comment:
It seems that `currentLeaderElectionSession.cancel(true)` will not cancel the future of `LeaderElector#start()`.

Reply:
true, I was wrong in my assumption that `cancel` would cancel also future chains that are created by `thenCompose`. But tbh, the `thenCompose` is not really necessary because we're cancelling the previous leadership session before re-re-initiating another session all being guarded by a lock. I fixed it and added another test case to cover this behavior. :+1: